core/src/main/java/org/infinispan/marshall/core/internal/ExternalJBossMarshaller.java

```diff
@@ -0,0 +1,226 @@
+import org.infinispan.commons.marshall.StreamingMarshaller;
+import org.infinispan.configuration.global.GlobalConfiguration;
+import org.infinispan.marshall.core.JBossMarshaller;
+import org.jboss.marshalling.ByteInput;
+import org.jboss.marshalling.ByteOutput;
+import org.jboss.marshalling.ObjectTable;
+import org.jboss.marshalling.Unmarshaller;
+
+final class ExternalJBossMarshaller implements StreamingMarshaller {
+
+   final MarshallableTypeHints marshallableTypeHints = new MarshallableTypeHints();
+   final JBossMarshaller marshaller;
+
+   public ExternalJBossMarshaller(InternalExternalizerTable externalizers, GlobalConfiguration globalCfg) {
+      this.marshaller = new InternalJBossMarshaller(externalizers, globalCfg);
+   }
+
+   @Override
+   public void objectToObjectStream(Object obj, ObjectOutput out) throws IOException {
+      assert ExternallyMarshallable.isAllowed(obj) : "Check support for: " + obj.getClass();
+
+      if (out instanceof PositionalBuffer.Output) {
+         PositionalBuffer.Output posOut = (PositionalBuffer.Output) out;
+
+         BufferSizePredictor sizePredictor = marshallableTypeHints
+               .getBufferSizePredictor(obj.getClass());
+         int estimatedSize = sizePredictor.nextSize(obj);
+
+         int beforePos = posOut.savePosition();
+
+         ObjectOutput jbossOut = marshaller.startObjectOutput(
+               new JBossByteOutput(out), false, estimatedSize);
+         try {
+            jbossOut.writeObject(obj);
+         } finally {
+            marshaller.finishObjectOutput(jbossOut);
+         }
+
+         int afterPos = posOut.writePosition(beforePos);
+         sizePredictor.recordSize(afterPos - beforePos);
+      } else {
+         throw new IOException("External JBoss Marshalling requires position tracking object output");
+      }
+   }
+
+   @Override
+   public Object objectFromObjectStream(ObjectInput in) throws IOException, ClassNotFoundException {
+      if (in instanceof PositionalBuffer.Input) {
+         PositionalBuffer.Input posIn = (PositionalBuffer.Input) in;
+
+         int pos = in.readInt();
+         ObjectInput jbossIn = marshaller.startObjectInput(new JBossByteInput(in), false);
+         try {
+            return jbossIn.readObject();
+         } finally {
+            marshaller.finishObjectInput(jbossIn);
+            posIn.rewindPosition(pos);
+         }
+      } else {
+         throw new IOException("External JBoss Marshalling requires position rewinding object output");
+      }
+   }
+
+   @Override
+   public void start() {
+      marshaller.start();
+   }
+
+   @Override
+   public void stop() {
+      marshaller.stop();
+   }
+
+   @Override
+   public boolean isMarshallable(Object o) throws Exception {
+      return marshaller.isMarshallable(o);
+   }
+
+   @Override
+   public ObjectOutput startObjectOutput(OutputStream os, boolean isReentrant, int estimatedSize) throws IOException {
+      throw new RuntimeException("NYI");
```

Comment:
Do you really need to implement these methods, or can you throw an `UnsupportedOperationException` instead?